cmd/kubelet/app/server.go

```diff
@@ -124,25 +129,79 @@ is checked every 20 seconds (also configurable with a flag).
 
 HTTP server: The kubelet can also listen for HTTP and respond to a simple API
 (underspec'd currently) to submit a new manifest.`,
+		// The Kubelet has special flag parsing requirements to enforce flag precedence rules,
+		// so we do all our parsing manually in Run, below.
+		// DisableFlagParsing=true provides the full set of flags passed to the kubelet in the
+		// `args` arg to Run, without Cobra's interference.
+		DisableFlagParsing: true,
 		Run: func(cmd *cobra.Command, args []string) {
+			// initial flag parse, since we disable cobra's flag parsing
+			if err := cleanFlagSet.Parse(args); err != nil {
+				cmd.Usage()
+				glog.Fatal(err)
+			}
+
+			// short-circuit on help
+			help, err := cleanFlagSet.GetBool("help")
+			if err != nil {
+				glog.Fatal(`"help" flag is non-bool, programmer error, please correct`)
+			}
+			if help {
+				cmd.Help()
+				return
+			}
+
 			// short-circuit on verflag
 			verflag.PrintAndExitIfRequested()
 
-			// TODO(mtaufen): won't need this this once dynamic config is GA
-			// set feature gates so we can check if dynamic config is enabled
-			if err := utilfeature.DefaultFeatureGate.SetFromMap(kubeletConfiguration.FeatureGates); err != nil {
+			// log args (separate lines, so we don't overflow line-length limits in logging infrastructure)
+			glog.V(2).Infof("kubelet flags: %s", strings.Join(args, "\n"))
+
+			// set feature gates from initial flags-based config
+			if err := utilfeature.DefaultFeatureGate.SetFromMap(kubeletConfig.FeatureGates); err != nil {
 				glog.Fatal(err)
 			}
-			// validate the initial KubeletFlags, to make sure the dynamic-config-related flags aren't used unless the feature gate is on
+
+			// validate the initial KubeletFlags
 			if err := options.ValidateKubeletFlags(kubeletFlags); err != nil {
 				glog.Fatal(err)
 			}
-			// bootstrap the kubelet config controller, app.BootstrapKubeletConfigController will check
-			// feature gates and only turn on relevant parts of the controller
-			kubeletConfig, kubeletConfigController, err := BootstrapKubeletConfigController(
-				kubeletConfiguration, kubeletFlags.KubeletConfigFile, kubeletFlags.DynamicConfigDir)
-			if err != nil {
-				glog.Fatal(err)
+
+			// load kubelet config file, if provided
+			if configFile := kubeletFlags.KubeletConfigFile; len(configFile) > 0 {
+				kubeletConfig, err = loadConfigFile(configFile)
```

Comment:
BUG: This clobbers `kubeletConfig` and the defaults populated by `applyLegacyDefaults` within `NewKubeletConfiguration`

https://github.com/kubernetes/kubernetes/blob/cc76a756e409d08959b8b6e623e317619399f015/cmd/kubelet/app/options/options.go#L242-L270

Reply:
That is intentional. The v1beta1 defaults get reapplied to what is read from the config file. When reading from a config file, we have *different* and better defaults than the command line.

Reply:
That's what the godoc on applyLegacyDefaults was trying to express (these are defaults that *only* apply to the CLI flags and do *not* apply when running kubelet from a config file)

Reply:
ah thanks for clearing that up jordan! surprised you responded xD hope all is well